python/ray/dag/nccl_operation.py

```diff
@@ -0,0 +1,31 @@
+from abc import ABC, abstractmethod
+from typing import List, Set, Any
+
+
+class _NcclOperation(ABC):
```

Comment:
It is a bit confusing that this class is used both at scheduling time (`is_ready`) and at execution time (`execute`). That also means that the same `_NcclOperation` cannot be used for two scheduling passes.

It would be better if we can have one immutable operation that is basically used just to specify the `execute` definition, and another operation that is created on-the-fly before scheduling. I think the latter one could also be reused directly for scheduling p2p operations.